sound/soc/sof/intel/hda-mlink.c

```diff
@@ -170,6 +170,69 @@ static int hdaml_lnk_enum(struct device *dev, struct hdac_ext2_link *h2link,
 	return 0;
 }
 
+/*
+ * Hardware recommendations are to wait ~10us before checking any hardware transition
+ * reported by bits changing status.
+ * This value does not need to be super-precise, a slack of 5us is perfectly acceptable.
+ * The worst-case is about 1ms before reporting an issue
+ */
+#define HDAML_POLL_DELAY_MIN_US 10
+#define HDAML_POLL_DELAY_SLACK_US 5
+#define HDAML_POLL_DELAY_RETRY  100
+
+static int check_power_active(u32 __iomem *lctl, int sublink, bool enable)
+{
+	int mask = BIT(sublink) << AZX_ML_LCTL_CPA_SHIFT;
+	int retry = HDAML_POLL_DELAY_RETRY;
+	u32 val;
+
+	usleep_range(HDAML_POLL_DELAY_MIN_US,
+		     HDAML_POLL_DELAY_MIN_US + HDAML_POLL_DELAY_SLACK_US);
+	do {
+		val = readl(lctl);
+		if (enable) {
+			if (val & mask)
+				return 0;
+		} else {
+			if (!(val & mask))
+				return 0;
+		}
+		usleep_range(HDAML_POLL_DELAY_MIN_US,
+			     HDAML_POLL_DELAY_MIN_US + HDAML_POLL_DELAY_SLACK_US);
+
+	} while (--retry);
+
+	return -EIO;
```

Comment:
humm, not sure. This part is supposed to be relatively simple so that the code can be ported easily to other platforms/environments. Indeed it's open-coding but it's also not supposed to be too Linux-specific. I wasn't able to figure out what the delay was here, no idea what (HDAML_POLL_DELAY_SLACK_US >> 2) + 1 comes from.

@ranj063 what would you prefer?

Reply:
@plbossart, it comes from:
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/linux/iopoll.h#n55

Reply:
@ujfalusi @ranj063 I will lean on the side of keeping the open-coding for now.

The main reason is that the sequences are similar to existing kernel code in sound/hda/ext and drivers/soundwire, but with the timing adjusted per hardware recommendations.

I think it's better in a first step to keep the code similar to previous functions, and in a second step, if desired, to use an optimized routine.

Reply:
no change.